src/controllers/secret.go

```diff
@@ -0,0 +1,194 @@
+	"go.uber.org/zap"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8stypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+)
+
+type SecretController interface {
+	CreateSecret(types.CreateSecretRequest) (types.CreateSecretResponse, error)
+	GetSecrets(types.GetSecretsRequest) (types.GetSecretsResponse, error)
+	GetSecret(types.GetSecretRequest) (types.GetSecretResponse, error)
+	PatcnSecret(types.UpdateSecretRequest) (types.UpdateSecretResponse, error)
+	DeleteSecret(types.DeleteSecretRequest) (types.DeleteSecretResponse, error)
+}
+
+type secretController struct {
+	client *kubernetes.Clientset
+	ctx    context.Context
+	logger *zap.Logger
+}
+
+func NewSecretController(client *kubernetes.Clientset, context context.Context, logger *zap.Logger) SecretController {
+	return &secretController{
+		client: client,
+		ctx:    context,
+		logger: logger,
+	}
+}
+
+func (n *secretController) CreateSecret(request types.CreateSecretRequest) (types.CreateSecretResponse, error) {
+	n.logger.Debug(fmt.Sprintf("Trying to create secret: %q", request.Name))
+
+	var secretType v1.SecretType
+	response := types.CreateSecretResponse{}
+	name := request.Name
+	namespace := request.Namespace
+	data := make(map[string][]byte)
+
+	if request.Type == "tls" {
+		if request.Cert == "" || request.Key == "" {
+			return response, fmt.Errorf("cert and key are required for TLS secrets")
+		}
+		secretType = v1.SecretTypeTLS
+		data["tls.crt"] = []byte(request.Cert)
+		data["tls.key"] = []byte(request.Key)
+	} else if request.Type == "opaque" {
+		secretType = v1.SecretTypeOpaque
+		for _, kv := range request.Data {
+			data[kv.Key] = []byte(kv.Value)
+		}
+		if len(data) == 0 {
+			return response, fmt.Errorf("data is required for Opaque secrets")
+		}
+	} else {
+		return response, fmt.Errorf("unsupported secret type: %s", request.Type)
+	}
+
+	secret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      request.Name,
+			Namespace: request.Namespace,
+		},
+		Type: secretType,
+		Data: data,
+	}
+
+	newSecret, err := n.client.CoreV1().Secrets(namespace).Create(n.ctx, secret, metav1.CreateOptions{})
+	if err != nil {
+		n.logger.Error(fmt.Sprintf("Could not create secret %q with error: %s", name, err.Error()))
+		return response, err
+	}
+
+	n.logger.Debug(fmt.Sprintf("Created secret %q successfully", name))
+
+	response.Type = string(newSecret.Type)
+	response.Name = newSecret.Name
+	response.Namespace = newSecret.Namespace
+	return response, err
+}
+
+func (n *secretController) GetSecrets(request types.GetSecretsRequest) (types.GetSecretsResponse, error) {
+	n.logger.Debug(fmt.Sprintf("Trying to get all secrets in %s namespace", request.Namespace))
+
+	response := types.GetSecretsResponse{}
+	namespace := request.Namespace
+	secrets, err := n.client.CoreV1().Secrets(namespace).List(n.ctx, metav1.ListOptions{})
+	if err != nil {
+		n.logger.Error(fmt.Sprintf("Could not get secrets with error: %s", err.Error()))
```

Comment:
We normally format the error with `%v` instead of `%s`. It doesn't make a huge difference but we'd like to keep it consistent in the code-base